x-pack/plugins/lens/public/editor_frame_service/embeddable/embeddable.tsx

```diff
@@ -128,6 +136,21 @@ export class Embeddable extends AbstractEmbeddable<LensEmbeddableInput, LensEmbe
     }
   }
 
+  async initializeSavedVis(input: LensEmbeddableInput) {
+    const attributes = await this.deps.attributeService.unwrapAttributes(input);
+    this.savedVis = {
+      ...attributes,
+      type: this.type,
+      savedObjectId: (input as LensByReferenceInput)?.savedObjectId,
+    };
+    const expression = await this.deps.documentToExpression(this.savedVis);
```

Comment:
Maybe you can simplify the whole thing by changing the signature of `documentToExpression`?